File: playwright-typescript/src/support/project-paths.ts
import fs from "node:fs";
import path from "node:path";

function findRepositoryRoot(): string {
  let candidate = process.cwd();
  while (true) {
    if (fs.existsSync(path.join(candidate, "manage.py"))) {
      return candidate;
    }
    const parent = path.dirname(candidate);
    if (parent === candidate) {
      throw new Error("Could not locate the Fictoshop repository root containing manage.py.");
    }
    candidate = parent;
  }
}

const root = findRepositoryRoot();

export const projectPaths = {
  root,
  python: path.join(root, ".venv", "bin", "python"),
  managePy: path.join(root, "manage.py")
};

File: playwright-csharp/Support/ProjectPaths.cs
namespace Fictoshop.PlaywrightTests.Support;

public static class ProjectPaths
{
    public static string Root { get; } = FindProjectRoot();

    public static string Python => Path.Combine(Root, ".venv", "bin", "python");

    private static string FindProjectRoot()
    {
        var candidates = new[] { Directory.GetCurrentDirectory(), AppContext.BaseDirectory };
        foreach (var candidate in candidates)
        {
            var directory = new DirectoryInfo(candidate);
            while (directory is not null)
            {
                if (File.Exists(Path.Combine(directory.FullName, "manage.py")))
                {
                    return directory.FullName;
                }

                directory = directory.Parent;
            }
        }

        throw new DirectoryNotFoundException("Could not locate the Fictoshop root containing manage.py.");
    }
}
